accepted/cranelift-roadmap-2022.md

```diff
@@ -0,0 +1,895 @@
+currently in a late stage, waiting for more implementations to come
+online.
+
+A tail call occurs when a call instruction is the last instruction in
+a function body, and the result of the call is directly returned. That
+call is then known as a "tail call" and can actually be replaced with
+a jump. This way, the tail-called function can return directly to our
+caller, saving the side-trip back through the current function. Once a
+language or compiler IR has tail calls, they can actually be used in
+many ways: for example, they are they primary control flow mechanism
+in Scheme and are used to implement looping. When present in Wasm,
+they become a very useful tool for compilers targeting Wasm bytecode.
+
+The difficulty with implementing tail-call support comes in the
+*uncommon cases* (as it does for so many compiler features). In other
+words: when everything goes right, a tail call is just a jump, after
+putting the arguments in the correct registers. But if the tail-called
+function has stack arguments, and if we pass some of them by loading
+values from our stack spillslots, and if we have to remove our
+stackframe before making the call (a requirement of proper tail-call
+support to ensure that an infinite sequence of tail-calls does not
+grow an unbounded stack nest), then we quickly have a general "state
+permutation" problem with difficult corner cases. [This GitHub
+comment](https://github.com/bytecodealliance/wasmtime/issues/1065#issuecomment-914420622)
+describes the problem in a little more detail.
+
+To implement this, we'll need to implement a new internal ABI in which
+callees clean up their stack frames; and then we'll need to implement
+the tail-call operator itself, including the corner-cases described
+above. The primary Wasm-engine consumer of Cranelift, Wasmtime,
+already uses trampolines to bridge between Wasm and host-ABI code, so
+a new ABI is not a large problem; getting the corner cases right is
+the hard part!
+
+Once we have tail-call support, we will likely want to look at related
+optimizations such as a
+[contification](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.83.3937&rep=rep1&type=pdf)
+pass, which when tail-call-aware is capable of converting
+recursion-based flow control into loops and other more efficient
+structures.
+
+### Exception Handling
+
+Another language feature that requires explicit compiler support is
+*exception handling*: the ability to unwind stack frames and perform
+cleanup along the way, stopping at relevant handlers, all without
+overhead in the common (non-exception-throwing) case.
+
+Zero-overhead exception support requires several things from the
+compiler. First, it must emit metadata that correlates locations in
+the program (PC values) to relevant handlers. It must also emit
+precise metadata that enables unwinding of stack frames from any point
+that might experience a thrown exception. Finally, exceptional control
+flow must be taken into account by optimizations.
```

Comment:
Also the landingpad must be able to take parameters. The x86_64 DWARF exception handling allows up to 4. Rustc requires 2.

Reply:
Added "and dataflow", thanks!